### ISSUE TITLE
Provide a way to limit HTTP/2 initial max concurrent streams by a client

### ISSUE THREAD

**Issue by minwoox** [line/armeria#3448]
A client can make more streams than the server allows and send them before it gets the HTTP/2 settings frame from the server.
We need a way to limit the number:
```java
ClientFactory.builder()
             .http2InitialMaxConcurrentStreams()
             /* .http2MaxConcurrentStreams() Will be added when server push is implemented */
             ...
```

**Comment by minwoox** [comment]
Speaking of max concurrent streams, I think the method name for setting the max current streams server-side can be mistakenly considered. (`Server.builder().http2MaxStreamsPerConnection`)
I hope we can change it to `http2MaxConcurrentStreams` when we reach 2.0.

**Comment by trustin** [comment]
> A client can make more streams than the server allows and send them before it gets the HTTP/2 settings frame from the server.

Shouldn't we wait until HTTP/2 SETTINGS frame is received so we know the MAX_CONCURRENT_STREAMS value the server suggests?

**Comment by trustin** [comment]
> .http2InitialMaxConcurrentStreams()

This is not really necessary for our case because we can just obey the max concurrent streams the server advertises in the initial HTTP/2 SETTINGS frame the server sends. We just need to make sure we don't put the connection that didn't receive the SETTINGS frame into the pool.

**Comment by minwoox** [comment]
> This is not really necessary for our case because we can just obey the max concurrent streams the server advertises in the initial HTTP/2 SETTINGS frame the server sends. 

Yeah, probably we don't need that because the client might not know before it gets the settings frame.

> We just need to make sure we don't put the connection that didn't receive the SETTINGS frame into the pool.

I'm a bit worried that we are going to make too many connections (>= 100 in this case), if we let the client send the request before it gets the settings frame. 🤔

**Comment by minwoox** [comment]
If we need to wait for the settings frame from the server, there might be a little delay for the first request.
If the client creates connections for each request before it gets the settings frame, there might be a chance that a lot of connections created.
So I'm a bit worried about both options.

I guess adding `http2InitialMaxConcurrentStreams` will not solve the problem, but it can be just an option for the client to change the default value which is `Integer.MAX_VALUE` now. 🤔

**Comment by anuraaga** [comment]
Isn't it illegal in HTTP/2 spec to send any data before reading the settings frame from the server?

Client's connection preface comes after server's settings frame
https://tools.ietf.org/html/rfc7540#section-3.2

**Comment by anuraaga** [comment]
Might be missing something, but it sounds like connections do need to be in the pool while sending SETTINGS, but all the client streams (concretely, sending connection preface) should be delayed until it comes in.

**Comment by trustin** [comment]
> Isn't it illegal in HTTP/2 spec to send any data before reading the settings frame from the server?

Correct. We have to wait before sending anything.

> If the client creates connections for each request before it gets the settings frame, there might be a chance that a lot of connections created.

We don't. We wait for the existing session establishment attempt in progress if there's one already. See `usePendingAcquisition`

**Comment by minwoox** [comment]
> Isn't it illegal in HTTP/2 spec to send any data before reading the settings frame from the server?

I guess it's not illegal.
https://tools.ietf.org/html/rfc7540#section-3.5
```
   To avoid unnecessary latency, clients are permitted to send
   additional frames to the server immediately after sending the client
   connection preface, without waiting to receive the server connection
   preface.  It is important to note, however, that the server
   connection preface SETTINGS frame might include parameters that
   necessarily alter how a client is expected to communicate with the
   server.
```

**Comment by minwoox** [comment]
> We don't. We wait for the existing session establishment attempt in progress if there's one already. See usePendingAcquisition

I misunderstood what you meant. 😅 
Yeah, that does not make an additional connection.
So the established channel is complete when it receives the settings frame from the server?

**Comment by anuraaga** [comment]
>    To avoid unnecessary latency, clients are permitted to send additional frames to the server immediately after sending the client connection preface

@minwoox - Ah, so the default behavior should definitely be to wait for the SETTINGS frame. Could add `.http2InitialMaxConcurrentStreams()` or just boolean `eagerlyStartClientStreams()` for users that know they won't cause problems by this eager behavior.

**Comment by minwoox** [comment]
> eagerlyStartClientStreams()

Yeah, that's a good suggestion. 👍  A user can make a choice for the initial behavior.
If a user enables that option, then we need the setting for the max concurrent streams as well?
`... eagerlyStartClientStreams(int http2InitialMaxConcurrentStreams) {...}`

**Comment by minwoox** [comment]
Or probably `Http2Settings` instead of `http2InitialMaxConcurrentStreams`? 🤔

**Comment by trustin** [comment]
What would a user specify in `Http2Settings` besides the max number of concurrent streams?

**Comment by minwoox** [comment]
> What would a user specify in Http2Settings besides the max number of concurrent streams?

Not so much? 🤣 
I thought there's a chance that a user wants to set the properties such as `maxFrameSize`, `initialWindowSize` and etc before the client gets the settings frame, but probably that's overkill.
